### PR TITLE
fix(client): reconnect default audio device on disconnect

### LINKED ISSUE #1092: [Quest 1] Unplugging headphones causes all audio to silence
## Description
Steps to reproduce:
- Open ALVR on Quest 1 (or quest 2?), connect it to a server on desktop, make sure an audio device is being routed to the headset, and then unplug headphones from the headset or plug them in

## General Troubleshooting
- [ x ] I carefully followed the instructions in the [README](https://github.com/alvr-org/ALVR/blob/master/README.md) and successfully completed the setup wizard
- [ x ] I read the ALVR Wikis [here](https://github.com/polygraphene/ALVR/wiki) and [here](https://github.com/alvr-org/ALVR/wiki)

## Software
Windows 10 Pro 21H2 build 19044.1766
SteamVR Beta 1.23.2
ALVR Server v18.2.3

### Hardware
AMD High Definition Audio Sound Card (?)
AMD Ryzen 5 2400G with Radeon Vega Graphics 3.60 GHz
Radeon RX 580 Series

**Install Type**:
- [ x ] Packaged (`exe`, `deb`, `rpm`, etc)

### PR DESCRIPTION
**Pull request by Okabintaro**
This makes it possible to plug in your headphones without restarting the client(Fixes #1092).

I am still new to rust and not sure if this is best way to solve this but it works for me and it annoyed me a bunch. Open for suggestions!

### CODE REVIEW

**Review by galister — CHANGES_REQUESTED** [review]
I would avoid adding an additional Mutex here. Instead, maybe put the `sample_buffer` and `state` together in a new struct, and make them share the existing `Arc<Mutex>`.

For example:

```rust
pub struct AudioLoopData<D> {
    sample_buffer: VecDeque<f32>,
    data: D
}
```

```rust
pub fn receive_samples_loop<D>(
    [...]
    data: Arc<Mutex<AudioLoopData<D>>>
```

**Comment by Okabintaro** [comment]
There are some reasons why I did it that way but I think it could be possible:

1. The callback that sets the state can potentially be called from yet another thread, which could result in a deadlock between the [on_audio_ready](https://docs.rs/oboe/latest/oboe/trait.AudioOutputCallback.html#tymethod.on_audio_ready) and the [on_error_before_close](https://docs.rs/oboe/latest/oboe/trait.AudioOutputCallback.html#method.on_error_before_close) if they would share a single mutex.

2. I was following the example of the `AudioPlayBackData` from above which also introduced a mutex for that state. So you could argue this is consistent: https://github.com/alvr-org/ALVR/blob/455d9d535f7565744cea3349077987c5670dbcc2/alvr/client_core/src/audio.rs#L16-L20

3. I also think having them separate is good because the state one will be locked quite often by the loop that puts in the data for the other callback to read: https://github.com/alvr-org/ALVR/blob/455d9d535f7565744cea3349077987c5670dbcc2/alvr/client_core/src/audio.rs#L191

Maybe one could use an atomic to make things more performant but it's working fine as it is at the moment.

You could workaround 1 probably, by using [on_error_after_close](https://docs.rs/oboe/latest/oboe/trait.AudioOutputCallback.html#method.on_error_after_close) which should get called in the same thread than the [on_audio_ready](https://docs.rs/oboe/latest/oboe/trait.AudioOutputCallback.html#tymethod.on_audio_ready).

File: alvr/client_core/src/audio.rs [hunk 1]
```diff
@@ -130,6 +136,14 @@ impl AudioOutputCallback for PlayerCallback {
 
         DataCallbackResult::Continue
     }
+
+    fn on_error_before_close(
+        &mut self,
+        _audio_stream: &mut dyn AudioOutputStreamSafe,
+        _error: oboe::Error,
```
Comment (zarik5) [inline alvr/client_core/src/audio.rs:143]: Remove leading underscore in `_error`

File: alvr/client_core/src/audio.rs [hunk 2]
```diff
@@ -184,5 +200,9 @@ pub fn play_audio_loop(
     // Note: Oboe crahes if stream.stop() is NOT called on AudioPlayer
     stream.stop_with_timeout(0).ok();
 
-    Ok(())
+    let result = match *state.lock() {
+        AudioPlaybackState::Err(error) => Err(error.into()),
+        _ => Ok(()),
+    };
```
Comment (zarik5) [inline alvr/client_core/src/audio.rs:206]: Since you are matching all variants of the enum, don't use wildcard `_`

**Review by zarik5 — CHANGES_REQUESTED** [review]
Overall LGTM!

**Comment by Okabintaro** [comment]
Thanks for the review!

I've been using this for quite some time on my Quest 2 but would maybe like someone else to test it too.
For some reason recently the quest microphone input just stops working and I can't even bring up the quest dashboard app to restart ALVR. I am not sure if this is just the Quest being buggy or these changes. After rebooting things work fine again.

**Review by galister — APPROVED** [review]

**Review by zarik5 — APPROVED** [review]

**Pull request merged**